cpp/src/arrow/types/primitive-test.cc

```diff
@@ -214,10 +214,10 @@ void TestPrimitiveBuilder<PBoolean>::Check(
     ex_null_bitmap = nullptr;
   }
 
-  auto expected = std::make_shared<BooleanArray>(size, ex_data, ex_null_count,
-      ex_null_bitmap);
-  std::shared_ptr<BooleanArray> result = std::dynamic_pointer_cast<BooleanArray>(
-      builder->Finish());
+  auto expected =
+      std::make_shared<BooleanArray>(size, ex_data, ex_null_count, ex_null_bitmap);
```

Comment:
I was wrong about the exact cause of this.  I turned on bin-packing and it didn't help much, I tried tweaking one or two other things, but it didn't seem to help.  Anyways, the latest changes are here on a new commit.